### PR TITLE
Ner

### PR DESCRIPTION
**Pull request by 200kgsquat**

### CODE REVIEW

File: docs/ner_predictions.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# BERT NER Predictions — Analysis
+
+This document summarizes predictions made by the BERT NER model (`bert-base-cased`) on the `jjzha/skillspan` dataset.
```
Comment (ernestknurov) [inline docs/ner_predictions.md:3]: I would like to know training process details e.g. on how many epochs model were trained, how much data was used. Also I would like to see what model was used for baseline and what metrics were achieved. It's good for comparison.

File: docs/ner_predictions.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+## 2. Sample Predictions
+
+**Example 1:**  
+Sentence: "The collaboration also covers acute patients during peak periods ."  
+Predicted labels: `O O O O O O O O O O`  
+True labels: `O O O O O O O O O O`  
+> Model correctly predicted all tokens as `O`.
+
+**Example 2:**  
+Sentence: "4 ) Work on a mix of front-end back-end and cloud technologies ."
```
Comment (ernestknurov) [inline docs/ner_predictions.md:45]: I don't think the described sentence has skills. Did you take this example from labeled dataset?

File: src/app/core/ner/bert_ner.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+            if self.use_focal and labels is not None:
+                num_labels = logits.shape[-1]
+                logits_flat = logits.view(-1, num_labels)
+                labels_flat = labels.view(-1)
+                ce_loss = F.cross_entropy(logits_flat, labels_flat, reduction='none', ignore_index=-100)
+                pt = torch.exp(-ce_loss)
+                focal_loss = self.focal_alpha * (1 - pt) ** self.focal_gamma * ce_loss
+                mask = labels_flat != -100
+                loss = focal_loss[mask].mean() if mask.sum() > 0 else torch.tensor(0.0, device=logits.device, requires_grad=True)
```
Comment (ernestknurov) [inline src/app/core/ner/bert_ner.py:76]: focal loss - good idea 👍

File: src/app/core/ner/bert_ner.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                return (loss, outputs) if return_outputs else loss
```
Comment (ernestknurov) [inline src/app/core/ner/bert_ner.py:77]: But why didn't you use the CRF with classification head? Without CRF you can get illegal or unlikely sequence of tags. Did you measure how often your model produce illegal sequence of tags?

**Review by ernestknurov — APPROVED** [review]
Please provide the PR description next time, e.g. what's changed.
Overall is good, but lacks the documentation details and CRF (or explanation that you don't need it)

File: docs/ner_predictions.md (hunk 2 above)
Comment (200kgsquat) [reply docs/ner_predictions.md:45]: yes, it was random sample from my data

File: src/app/core/ner/bert_ner.py (hunk 4 above)
Comment (200kgsquat) [reply src/app/core/ner/bert_ner.py:77]: If I get mostly illegal sequences, it will affect my metrics, for instance F1 span. But it doesn’t, so I can say BERT takes over CRF’s work.

**Review by ernestknurov — APPROVED** [review]
👍
